feat(gui): Toolbar context menu to display the buttons as or with text

s4moore opened the pull request:
Added a context menu when right clicking on the tool tray which gives the option to customize the dsiplay of the icons by choosing between text, icon, text alongside icon, and text below icon

Close #1105

s4moore committed:
added context menu for selecting style of icons in tool tray

s4moore committed:
removed two cases of trailing spaces

buhtz commented:
Great. I will look later deeper into it. Do you know the application "peek"? With that you can create screencast, GIF-animated Screenshots of the Gui. 

Here is an example:
![Peek 2025-01-03 11-09](https://github.com/user-attachments/assets/db79c795-faca-4634-a6ca-0ff0af258cd6)

buhtz commented on qt/app.py:
```diff
@@ -735,12 +735,57 @@ def _create_menubar(self):
         restore.insertSeparator(self.act_restore_parent)
         restore.setToolTipsVisible(True)
 
+    def wrap_text(self, text, max_len):
```
There is no need for this function. Python offers the `textwrap` package for this. Have look into `qttools.set_wrapped_tooltip()` how we are using it.

s4moore replied:
I couldn't get this working properly but have used the TextWrapper class from the provided textwrap.py and removed the custom function. I also couldn't get peek working as I'm using WSL (been trying to solve for a couple of hours now :/ )

buhtz replied:
You can run Back In Time with Microsoft Windows WSL? Just the command line version "backintime" or "backintime-qt", too?

Anyway: What do you mean by "not properly". Please describe the goal you want to achieve with text wrapping. I might be able to help.

s4moore replied:
I have been able to get backintime-qt working but do get the following errors:

WARNING: Failed to connect to Udev serviceHelper daemon via D-Bus: org.freedesktop.DBus.Error.Spawn.ChildExited
WARNING: D-Bus message: Launch helper exited with unknown return code 1
WARNING: Udev-based profiles cannot be changed or checked due to Udev serviceHelper connection failure

I've looked into it a bit and it seems to be because WSL doesn't interact with hardware in the same way as native Linux but the program still does backups and seems okay otherwise

buhtz replied:
Wow. Good to know. But as a maintainer I would not recommend doing it that way.

buhtz commented on qt/app.py:
```diff
@@ -735,12 +735,57 @@ def _create_menubar(self):
+        """
+        Create a context menu for changing the style of the icons
+        """
+        context_menu = QMenu(self)
```
I would like to have an indicator (e.g. checkmark) in the context menu. This way it would be obvious to the user which option is currently active.

e.g. something like this

![image](https://github.com/user-attachments/assets/0bb25b01-1b61-42b2-8529-ed82266fa42b)

s4moore replied:
done, let me know if you're happy with the ✔ or prefer something else

buhtz reviewed (changes requested):
Hello Samuel,
great work. Thank you for this. The text wrapping is a good idea.

I added two comments suggesting some modifications.

s4moore committed:
added check for selected style and replaced custom line-wrap with qttools textwrapper

s4moore commented:
Just realised I said in my commit message that the textwrapper was from qttools but I think that's a mistake..

Also, I added the _() around the icon text as it seems that this is how you mark the text for translation?

s4moore committed:
cosmetic: update to use convenience function instead of TextWrapper class

buhtz commented on qt/app.py:
```diff
@@ -32,7 +32,7 @@
 
 import qttools
-from textwrap import TextWrapper
+from textwrap import fill
```
```suggestion
import textwrap
```

buhtz commented on qt/app.py:
```diff
@@ -799,7 +799,7 @@ def _create_main_toolbar(self):
                 toolbar.widgetForAction(act).setToolTip(button_tip)
 
-            act.setText(TextWrapper(width=8, break_long_words=False).fill(act.text()))
+            act.setText(fill(act.text(), width=8, break_long_words=False))
```
The term `fill` is to broad. I would like to keep the package name to make the context clear to the readers.
```suggestion
            act.setText(textwrap.fill(act.text(), width=8, break_long_words=False))
```

buhtz commented:
> Also, I added the _() around the icon text as it seems that this is how you mark the text for translation?

Yes, this is correct.

s4moore committed:
Import textwrap module instead of just the fill function

Importing the module rather than just the fill function so use of the the fill function is more explicitly part the the textwrap module

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Call textwrap.fill instead of fill

Calling textwrap.fill makes the purpose of the function call more explicit

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

buhtz commented on qt/app.py:
```diff
@@ -735,12 +736,34 @@ def _create_menubar(self):
+            (_('Text Beside Icon'), Qt.ToolButtonStyle.ToolButtonTextBesideIcon),
+        )
+        for text, style in options:
+            text = '✔ ' + text if toolbar.toolButtonStyle() == style else '  ' + text
```
I have to 😄 . I like this solution because it is so simple and easy. Very pythonic.
But I am assuming the visual appearance is unusual regarding the experience of our users.

I had something like the `QActionGroup` in mind.
![image](https://github.com/user-attachments/assets/590f7a14-aef2-4a42-bf86-1c97b9799694)

See [PyQt6 docu](https://www.riverbankcomputing.com/static/Docs/PyQt6/api/qtgui/qactiongroup.html#QActionGroup) (without pictures) and original [C++ Qt docu](https://doc.qt.io/qt-5/qactiongroup.html) (with pictures).

s4moore replied:
Okay, so I should implement the action group right? Regarding the visual appearance, do you mean of the code or the execution? Sorry if the question sounds strange, I'm just not entirely certain what would be best here?

buhtz replied:
Not strange. Don't worry. 😄 I mean the GUI. The code es beautiful. And it will be a bit more uglier when using QActionGroup. But the GUI will look "better" and more usual to our users. From a coding perspective I really like your solution.

s4moore replied:
Okay, I've implemented the QActionGroup (and pushed with [skip ci]). The default icon is a bullet point, should I change to radio buttons or what is your preference?

s4moore replied:
And thanks for the kind words 😄

buhtz commented on qt/app.py:
```diff
@@ -735,12 +736,34 @@ def _create_menubar(self):
+        """
+        context_menu = QMenu(self)
+        options = (
+            (_('Icon Only'), Qt.ToolButtonStyle.ToolButtonIconOnly),
+            (_('Text Only'), Qt.ToolButtonStyle.ToolButtonTextOnly),
+            (_('Text Under Icon'), Qt.ToolButtonStyle.ToolButtonTextUnderIcon),
+            (_('Text Beside Icon'), Qt.ToolButtonStyle.ToolButtonTextBesideIcon),
```
I am not a native speaker but this sounds easier to me. What do you think?

```suggestion
            (_('Icons only'), Qt.ToolButtonStyle.ToolButtonIconOnly),
            (_('Text only'), Qt.ToolButtonStyle.ToolButtonTextOnly),
            (_('Text below icons'), Qt.ToolButtonStyle.ToolButtonTextUnderIcon),
            (_('Text beside icon'), Qt.ToolButtonStyle.ToolButtonTextBesideIcon),
```

s4moore replied:
Yes, it reads much better

s4moore committed:
Format icon text fields for better readability

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

buhtz commented:
Just a tip. You can prevent TravisCI from running if you add "[skip ci]" at the end of your commit message. So if you know one commit won't be your last or final one you can skip the CI.

The background is everytime you push your commits to the PR the TravisCI starts running. But the "credits" on this machine are limited for us but I have no elegant solution to switch it of. And there is also the CO2 the machine produces.

s4moore committed:
implemented QActionGroup for context menu [skip ci]

buhtz committed:
minor mods, license info, changelog

buhtz reviewed (approved):
Congratulations! 🥳  It now looks fine to me and ready to merge. It is our project policy to cool down some days before merging a PR.

- I changed the name of  the method and also made it "private" (using the `_` as prefix).
- I added your nickname in the SPDX header (see line 6 in app.py).
- I added a changelog entry also with your nickname.
- Feel free to add your real name and email (not recommended because of spam) there.
- The styling of the radio buttons is OK that way. I am assuming this depends on the desktop theme used. I am not aware of another application as an example but I hope it is consistent.

Don't forget to update your local branch after my modifications.

A next step (in another PR) would be to make this toolbar style persistent between sessions. But please wait until #1989 is merged. Then you can use the new introduce state file to solve this.

Please let me know if you want to take this next step. But there is no hurry.

s4moore commented:
It's really exciting that my name will be published, thank you so much! I wanted to change my name in app.py to "Samuel Moore" but not sure if pushing that change from my side would require an unneccesary run of TravisCI?

More than happy to change the style of buttons if you do decide that you would prefer something different, just let me know.

And yes, I will be happy to work on next step 😄

buhtz commented:
> I wanted to change my name in app.py to "Samuel Moore" but not sure if pushing that change from my side would require an unneccesary run of TravisCI?

Just make sure to do `git fetch --prune --all && git pull`. This will update your local branch with the latest commit I have done. Then you are free to modify the name or anything else and commit (plus `[skip ci]`) and push again into the PR.

> And yes, I will be happy to work on next step 😄

Than just watch #1989 . You can start when it is merged.

s4moore committed:
update name from github handle [skip ci]

buhtz committed:
Merge branch 'dev' into toolbar_style_options

buhtz commented:
Merged #1989

s4moore commented:
Hi Christian, 

I had some issues updating to the latest version of BIT. When I first tried to use the updated version, ``backintime.json`` was not found. I toucuhed the file to create it which got me further but then I was getting key errors because some of the fields are not present in the template used in ``statedata.py 30 -> 45`` . 

Am I missing a something in set up?

buhtz commented:
Are you sure that you have the latest "dev" version?
The missing backintime.json should be a WARNING on the first run, without interrupting BIT. BIT then will auto-migrate values from the config file into a new backintime.json.

But even if the backintime.json is empty there shouldn't be a keyerror.

Mhm... I need to think about that.

Please try to reproduce the problem and report the terminal output in a new issue. Delete your backintime.json and show me the error. Touch the backintime.json and show me the error.

s4moore commented:
Hi Christian,

I have completed a first attempt at saving the icon styling in state data. Should I simply open a pull request?

buhtz commented:
Yes, present your work in a pull request please.
